Add Dockerfile

Raimo33 opened the pull request:
This PR adds a general purpose lightweight Dockerfile to easily build and run a Knots node.
The Dockerfile is meant to be run on the same host as were it is built from; this is because of  CPU architecture specific compatibility issues that not even Docker addresses.

Inspired by https://github.com/bitcoinknots/bitcoin/pull/170

Raimo33 commented:
I understand that with the new knots release we won't use `./configure`, and I'll update the Dockerfile to use `cmake` as soon as that is merged

electricalgrade commented:
Can switch your github to the latest tag?
That way you will get cmake.

electricalgrade commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,71 @@
+
+EXPOSE 8333 8332
+WORKDIR /var/lib/knots
+
```
Create a non-root User and switch to this user

Raimo33 replied:
it's almost impossible without strange tweaks to run a docker container with a non root user if that container needs to bind a writeable volume. I haven't managed to find a solution. feel free to contribute if you have a clean way to do so.

electricalgrade replied:
This is challenge with alpine. Hence i use debain. the image size is hardly 60Mb more. 
But all such quirky problems are avoided. Debian works fine with writable volumes + non-root users.
I have tested with macos, arm64 and x86. Also debain has first class support to glibc. Alpine uses musl which can create some issues. 

In my homelab i run a docker swarm and k8s cluster. Most of my images are debian and only for small containers like nginx etc i use alpine.

Ultimately you dont want to run docker as a root.

luke-jr commented:
Best to rebase this on top of 8cb6ab0b971a27ba255ad6a48959a8c7b84c00f3

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,71 @@
+    coreutils \
+    binutils
+
+ARG MAJOR_VERSION=28
+ARG VERSION=28.1.knots20250305
+ARG BASE_URL=https://bitcoinknots.org/files/${MAJOR_VERSION}.x/${VERSION}
+ARG TARBALL=bitcoin-${VERSION}.tar.gz
```
None of this is (or should be?) used?

Raimo33 replied:
yes, forgot to delete.

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,71 @@
+ARG CXXFLAGS="-O2 -march=native -pipe -flto"
+ARG LDFLAGS="-flto"
+
+RUN ./configure \
+    --quiet \
+    --bindir=/usr/local/bin/knots \
+    --disable-dependency-tracking \
+    --disable-tests \
+    --disable-gui-tests \
+    --disable-bench \
+    --disable-fuzz-binary \
+    --disable-man \
+    --disable-shared \
+    --disable-static \
+    --enable-wallet \
+    --without-bdb \
+    --with-sqlite \
+    --with-gui=no
```
Outdated for 29.x

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,71 @@
+
+COPY --from=builder /usr/local/bin/knots /usr/local/bin/knots
+
+# volume for the blockchain data
+VOLUME ["/var/lib/knots"]
+# bind mount for the configuration file
+VOLUME ["/etc/knots/bitcoin.conf"]
```
This is usually /var/lib/bitcoin and /etc/bitcoin/

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,71 @@
+# bind mount for the configuration file
+VOLUME ["/etc/knots/bitcoin.conf"]
+
+EXPOSE 8333 8332
```
Without configuration, 8332 is localhost only

Raimo33 commented:
you want me to remove the config bind mount @luke-jr ? why?

and for port 8332, I think it should still be documented and exposed even if for localhost only. self hosted wallets need it.

luke-jr commented:
>you want me to remove the config bind mount @luke-jr ?

no?

>and for port 8332, I think it should still be documented and exposed even if for localhost only. self hosted wallets need it.

should it maybe be configured to allow access from the host? is it possible to expose it *only* to the host?

Raimo33 commented:
> should it maybe be configured to allow access from the host? is it possible to expose it _only_ to the host?

access can be restricted to be only from the host, that can be done when running the container, not from the Dockerfile.
The `EXPOSE` keyword on the Dockerfile is just documentation; no actual exposing happens at build time.

This is what can be done at runtime:
```bash
docker run -d \
  --name bitcoinknots \
  -p 8333:8333 -p 127.0.0.1:8332:8332 \
  -v path/to/data:/var/lib/bitcoin \
  -v path/to/conf:/etc/bitcoin/bitcoin.conf \
  bitcoinknots:latest
  ```

luke-jr commented:
Can we forbid `latest` and force the user to choose a version?

Raimo33 commented:
> 

mhhh, I'm currently copying the source as is via `COPY . .`

if the user clones master, then it will be master. if the user clones a stable release, then it will be that. If you want me to write the dockerfile so that the user dynamically chooses which knots version to use, I should probably download tarballs from the release mirror. but that becomes complicated because i need to know the host arch.

Or maybe I can play with git branches

luke-jr commented:
Nah, that's fine too - I was just confused by the `latest` bit

Raimo33 commented:
Image builds fine, it's ready for testing. I don't have a synced blockchain at disposal, if anyone can please run the container and bind the `.bitcoin/` as volume to test if peers connect and bitcoind is reachable via RPC.

electricalgrade commented:
Bitcoind is causing segementation fault. Tried this on mac.

you can test using
docker run --rm -it --entrypoint /bin/sh bitcoinknots

and launch bitcoind --help

electricalgrade commented:
I find that this option -march=native causes failure on mac with docker.
without that the bitcoind runs..

If possible pls test it on other platforms.

you can also add health check for rpc port
HEALTHCHECK --interval=30s --timeout=5s --start-period=30s --retries=3 \
  CMD nc -z 127.0.0.1 8332 && \
      bitcoin-cli -datadir=/bitcoin/.bitcoin getblockchaininfo || exit 1
  
  also please compile bitcoin-cli
  
  If possible get the non-root user working.

Raimo33 commented:
I've added HEALTHCHECK and `--init` (tini). It runs correctly without segfault on my amd64 CPU. Peers connect successfully, blocks get synced.

<img width="920" height="373" alt="Screenshot from 2025-08-29 13-09-32" src="https://github.com/user-attachments/assets/9700c111-b193-48b0-bd51-2b411d6b356f" />
<img width="969" height="518" alt="Screenshot from 2025-08-29 13-09-51" src="https://github.com/user-attachments/assets/70be7dec-41c5-496e-8c99-3ee7d5bdd11a" />


I haven't managed to add the non-root user, even with debian. It's an issue I have struggled with for years. Basically the UID/GID inside the container need to match the UID/GID of the directories mounted on the host. There are a couple of workarounds possible:

1. set the UID/GID inside the container to 1000:1000 as most of the times the first non-root user has that id on unix.
2. let the user pass the UID/GID as compile time ARGs to the Dockerfile.

I think both are bad. the first approach ([which is used in mempool.space for example](https://github.com/mempool/mempool/tree/master/docker)) is not guaranteed to be always correct. The second approach is bad UX for the user.

This said, if you managed to find a solution to add a `USER` without volume permission issues please share it.
I suggest interacting with the LLM at [docs.docker.com](docs.docker.com)

Raimo33 commented:
Also, why are we disabling ZeroMQ?

electricalgrade commented:
Thanks.
Please note -march=native  will fail on mac.
root user is a non-start for docker containers. So we have to live the pains of non-root user.

electricalgrade commented:
> Also, why are we disabling ZeroMQ?

since wallet is enabled, zeromq should be enabled.

Raimo33 commented:
- added zeromq
- will test on arm64 (mac) today
- I'm debated on alpine vs debian -> musl vs glibc (opinions needed)
- I genuinely don't know how to make a non-root user work (implementation ideas needed)

electricalgrade commented:
Thanks.
pls note I had to remove bin in the CMAKE_INSTALL_PREFIX
-DCMAKE_INSTALL_PREFIX="/opt/bitcoin/build/"

without that the STRIP comand would fail.
if you get march=native for mac then please document the steps. On my mac docker (i use docker desktop and not colima in mac) i need to disable this.

Also for root/user , just start with user UID/GIO set 1000 and that can be overridden from commandline.

ARG BITCOIN_UID=1000
ARG BITCOIN_GID=1000
#Default to the bitcoin user, but allow override via `--user`
USER bitcoin

so if users have issue then can just override with --user root. This the non-root user will work for most cases and in cases where it does not then it can be overridden.

Also alpine is building and running so no need to switch to debain.

Raimo33 commented:
the switch to debian was more a performance concern. musl can be a lot slower than glibc. it doesn't have architecture specific optimizations. glibc on alpine would be the best.

Raimo33 commented:
- Fixed BINDIR
- Removed `march=native` as it not that useful with musl, considering it doesn't have architecture specific code and was causing segfault (likely invalid instruction) on arm64 targets.
- Added non-root user with instructions in the README.md

electricalgrade reviewed (approved):
This works.

luke-jr commented:
>Removed march=native as it not that useful with musl, considering it doesn't have architecture specific code

march is a compiler thing, not libc-related...

Raimo33 commented:
> march is a compiler thing, not libc-related...

right, I had a misconception regarding how GCC compiled code.

electricalgrade commented:
Any other pending items in this? If not can we get this merged for next release?

luke-jr commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+.git/
```
Will this break `cmake/script/GenerateBuildInfo.cmake`?

Raimo33 replied:
correct. removed .dockerignore entirely. Docker won't be able to cache the `COPY . .` layer but it's not a big deal.

luke-jr commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+.git/
+contrib/docker/
```
Any reason not to end with a newline?

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+WORKDIR /opt/bitcoin/build
+
+RUN cmake .. \
+    -DCMAKE_INSTALL_PREFIX="/opt/bitcoin/build/" \
```
Why would we be installing into the build dir?

Raimo33 replied:
to make the strip command less scary. changed to the more conventional /usr/local. no big deal.

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+    -DBUILD_DAEMON="ON" \
+    -DBUILD_CLI="ON" \
+    -DENABLE_WALLET="ON" \
+    -DWITH_ZMQ="ON" \
```
How do users access the zeromq sockets?

Raimo33 replied:
Users can configure ZeroMQ ports at runtime by setting the appropriate `zmqpub*` options in `bitcoin.conf`. The `EXPOSE` directive in the Dockerfile is purely informational and does not publish any ports. To access the ZeroMQ sockets from outside the container, users must explicitly map the container ports to the host using `-p host_port:container_port`. For example, if `bitcoin.conf` contains `zmqpubblock=tcp://0.0.0.0:1234`, the container should be started with `-p 1234:1234`.

Raimo33 replied:
I added a disclaimer on the README.md

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+    -DENABLE_WALLET="ON" \
+    -DWITH_ZMQ="ON" \
+    -DBUILD_GUI="OFF" \
+    -DBUILD_TESTS="OFF" \
```
Can we build the tests and run them at build time at least?

Raimo33 replied:
yes we can. great idea. added. increases build time by 7 minutes but worth it.

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+    -DWITH_MULTIPROCESS="OFF" \
+    -DINSTALL_MAN="OFF" \
+    -DWITH_CCACHE="OFF" \
+    -DAPPEND_CXXFLAGS="-O2" \
```
Any reason not to let the CMake defaults handle this?

Raimo33 replied:
enabling `CCACHE` makes no sense, the default is ON; but a Dockerfile that builds only once does not benefit from the cache. correct me if I'm wrong.

`INSTALL_MAN` makes very little sense, the default is `OFF`; but still, I think it's important to override it for future compatibility. It does not make sense today, it will not make sense tomorrow.

`WITH_MULTIPROCESS` can be removed. I'll let this be handled by CMake defaults.

`APPEND_CXXFLAGS` the defaults don't do any type of optimization. I figured why not. but feel free to remove it, no big deal.

luke-jr replied:
The default *should* do optimization... We should figure out why it isn't, if not!

luke-jr replied:
Without any cmake parameters, I see it using `-O2`...

Raimo33 replied:
my bad, will remove. good catch. just remember that with architecture specific optimizations such as `-O2` `-march=native` etc. the image isn't portable and only works on the host. If one day you want to upload an official image to [dockerhub](https://hub.docker.com/) it needs to be refactored completely.

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+RUN addgroup -S -g ${GROUP_ID} bitcoin && \
+    adduser -S -u ${USER_ID} -G bitcoin -H -s /bin/false bitcoin
+
+VOLUME ["/var/lib/bitcoin", "/etc/bitcoin/bitcoin.conf"]
```
Sorry, my fault: the init scripts use /var/lib/bitcoin**d**

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+ENTRYPOINT ["bitcoind", "-conf=/etc/bitcoin/bitcoin.conf", "-datadir=/var/lib/bitcoin"]
+HEALTHCHECK --interval=5m --timeout=15s --start-period=2m --start-interval=10s \
+    CMD ["bitcoin-cli", "-conf=/etc/bitcoin/bitcoin.conf", "-datadir=/var/lib/bitcoin", "getblockchaininfo"]
```
Ends w/o a newline?

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+
+ENTRYPOINT ["bitcoind", "-conf=/etc/bitcoin/bitcoin.conf", "-datadir=/var/lib/bitcoin"]
+HEALTHCHECK --interval=5m --timeout=15s --start-period=2m --start-interval=10s \
+    CMD ["bitcoin-cli", "-conf=/etc/bitcoin/bitcoin.conf", "-datadir=/var/lib/bitcoin", "getblockchaininfo"]
```
How about symlinking /home/bitcoin/.bitcoin/bitcoin.conf to /etc/bitcoin/bitcoin.conf so bitcoin-cli picks it up automatically?

Raimo33 replied:
I personally don't like this. the bitcoin user is created without a home so `/home/bitcoin` doesn't even exist now. Feel free to change whatever you want though. @electricalgrade opinions?

luke-jr commented on contrib/docker/Dockerfile:
```diff
@@ -0,0 +1,67 @@
+RUN addgroup -S -g ${GROUP_ID} bitcoin && \
+    adduser -S -u ${USER_ID} -G bitcoin -H -s /bin/false bitcoin
+
+VOLUME ["/var/lib/bitcoin", "/etc/bitcoin/bitcoin.conf"]
```
What about bitcoin_rw.conf, settings.json, debug.log, etc?

Raimo33 replied:
wasn't aware of them...

@electricalgrade what do you think?

electricalgrade replied:
yes.  bitcoin_rw.conf should be provided. 
In my case settings.json is generated in the mounted directory. Same for debug.log

Everything is in /var/lib/bitcoin including banlist etc. It also needs /var/lib/bitcoin/fee_estimates.dat
Right now it does not find and just ignores them. 
The way i use it is that I have these fee_estimates.dat, bitcoin_rw.conf and banlist in folder which is mounted on /var/lib/bitcoin.
This way it finds the files needed.

I can view debug.log inside docker as well as it will be available on host. settings.json is autogenerated and stored in same directory.

ls ~/bitcoin-knots-data 
anchors.dat		bitcoin_rw.conf		debug.log		mempool.dat
banlist.json		blocks			fee_estimates.dat	peers.dat
bitcoin.conf		chainstate		indexes			settings.json

this is the run.sh 
docker run -d \
  --name bitcoinknots \
  -p 8333:8333 -p 127.0.0.1:8332:8332 \
  -v $HOME/bitcoin-knots-data:/var/lib/bitcoin \
  -v $HOME/bitcoin-knots-data/bitcoin.conf:/etc/bitcoin/bitcoin.conf \
  bitcoinknots

Raimo33 replied:
> yes. bitcoin_rw.conf should be provided.

I thought I knew what bitcoin_rw was, but turns out LLMs were blatantly lying to me.

Raimo33 replied:
> The way i use it is that I have these fee_estimates.dat, bitcoin_rw.conf and banlist in folder which is mounted on /var/lib/bitcoin.
This way it finds the files needed.

Yes, this is the only way. mounted Docker volumes override anything that's in the mapped directory. Host > Container. Do you think we should add a section in the readme to recommend pre-populating the data directory with `peers.dat`, `fee_estimates.dat` etc.?

Raimo33 replied:
Ok so no need to worry about `settings.json`, `debug.log` as they're available in the mapped `/var/lib/bitcoind`.

luke-jr replied:
But settings.json should be in /etc/bitcoin/ and debug.log probably in /var/log somewhere? OTOH, I think this is not how the init scripts work either, so maybe fine...

Raimo33 replied:
all these are created in /var/lib/bitcoind, including `bitcoin_rw.conf`. So the user has them mapped in their data volume. No need to complicate things further imo.

electricalgrade commented:
Will this be merged for v29? or are we late.
Any pending items?

Raimo33 commented:
@luke-jr I believe this is ready for merge
